File: Java/src/com/bh/no1071_Greatest_Common_Divisor_of_Strings.java
package com.bh;

public class no1071_Greatest_Common_Divisor_of_Strings {
    public String gcdOfStrings(String str1, String str2) {
        // If concatenating in both orders is not equal, no common divisor string exists
        if (!(str1 + str2).equals(str2 + str1)) {
            return "";
        }
        // Helper to compute GCD of two numbers
        int gcdLen = gcd(str1.length(), str2.length());
        return str1.substring(0, gcdLen);
    }
    // Euclidean algorithm for GCD
    private int gcd(int a, int b) {
        while (b != 0) {
            int temp = b;
            b = a % b;
            a = temp;
        }
        return a;
    }

    public static void main(String[] args) {
        no1071_Greatest_Common_Divisor_of_Strings solution = new no1071_Greatest_Common_Divisor_of_Strings();
        // Test case 1
        String str1_1 = "ABCABC";
        String str2_1 = "ABC";
        System.out.println("Test case 1: " + solution.gcdOfStrings(str1_1, str2_1)); // Expected: "ABC"
        // Test case 2
        String str1_2 = "ABABAB";
        String str2_2 = "ABAB";
        System.out.println("Test case 2: " + solution.gcdOfStrings(str1_2, str2_2)); // Expected: "AB"
        // Test case 3
        String str1_3 = "LEET";
        String str2_3 = "CODE";
        System.out.println("Test case 3: " + solution.gcdOfStrings(str1_3, str2_3)); // Expected: ""
    }
}

File: no1071_Greatest_Common_Divisor_of_Strings.py
"""
1071. Greatest Common Divisor of Strings

Given two strings str1 and str2, return the largest string x such that
both str1 and str2 are made of one or more concatenations of x.

Idea:
- If str1 + str2 != str2 + str1, no such base string exists -> return "".
- Otherwise, the answer length is gcd(len(str1), len(str2)); return str1[:g].

Time: O(n) to check concatenations, Space: O(1) extra.
"""

from math import gcd


class Solution:
	def gcdOfStrings(self, str1: str, str2: str) -> str:
		# If they don't share the same repeating pattern, no gcd string exists.
		if str1 + str2 != str2 + str1:
			return ""
		# Length of the gcd string is gcd of lengths.
		g = gcd(len(str1), len(str2))
		return str1[:g]


if __name__ == "__main__":
	tests = [
		("ABCABC", "ABC", "ABC"),
		("ABABAB", "ABAB", "AB"),
		("LEET", "CODE", ""),
		("AAAAAA", "AAA", "AAA"),
		("ABCABCABC", "ABCABC", "ABC"),
		("TAUXXTAUXXTAUXX", "TAUXXTAUXX", "TAUXX"),
		("ABAB", "ABBA", ""),
	]
	s = Solution()
	for i, (a, b, expected) in enumerate(tests, 1):
		got = s.gcdOfStrings(a, b)
		print(f"Case {i}: got='{got}', expected='{expected}'")


